File: e2e/src/ez_cdc_e2e/datasources/loader.py
"""YAML loader and variable interpolation for datasources files.

Pipeline:
  1. Read raw YAML text from disk.
  2. Optionally load `.env` files into `os.environ` (for `${VAR}` resolution).
  3. Walk every string value in the parsed YAML and substitute `${VAR}` with
     the corresponding env var. Missing vars raise InterpolationError unless
     a default is given via `${VAR:-default}`.
  4. Hand the substituted dict to pydantic for schema validation.

Errors are wrapped in DatasourceError subclasses with messages a CLI user
can act on (file path, line number when available, field hierarchy from
pydantic).

Variable interpolation syntax:
  ${VAR}              required — raises if VAR is not in environment
  ${VAR:-default}     optional — uses 'default' if VAR is unset or empty
  $$VAR or $${VAR}    literal $ — escape mechanism

This is intentionally minimal: no command substitution, no shell expansion,
no nested expansions. The YAML stays predictable.
"""

from __future__ import annotations

import os
import re
from pathlib import Path
from typing import Any

import yaml
from pydantic import ValidationError

from .schema import DatasourcesFile


# ── Exceptions ───────────────────────────────────────────────────────────────

class DatasourceError(Exception):
    """Base class for all datasource-related errors."""


class DatasourceNotFoundError(DatasourceError):
    """A named datasource was requested but doesn't exist in the file."""


class DatasourceValidationError(DatasourceError):
    """The YAML loaded but failed schema validation."""


class InterpolationError(DatasourceError):
    """A `${VAR}` reference couldn't be resolved."""


# ── Variable interpolation ───────────────────────────────────────────────────

# Match either ${VAR}, ${VAR:-default}, or the escape $${...} / $$VAR.
# We capture the optional default after `:-` so we can apply it.
_VAR_PATTERN = re.compile(
    r"""
    \$\$ (?: \{[^}]*\} | [A-Za-z_][A-Za-z0-9_]* )    # escape: $${VAR} or $$VAR
    |
    \$\{ (?P<name>[A-Za-z_][A-Za-z0-9_]*) (?: :- (?P<default>[^}]*) )? \}
    |
    \$ (?P<bare> [A-Za-z_][A-Za-z0-9_]* )            # bare $VAR (also supported)
    """,
    re.VERBOSE,
)


def interpolate_string(value: str, source_path: Path | None = None) -> str:
    """Substitute ${VAR} placeholders in a string.

    Args:
        value: the string to interpolate.
        source_path: file path used for error messages, if known.

    Raises:
        InterpolationError: when a referenced variable is not in os.environ
            and no default was supplied.
    """
    if "$" not in value:
        return value  # fast path

    def _replace(match: re.Match) -> str:
        full = match.group(0)
        # Escape sequence: $${...} or $$VAR — keep one literal $
        if full.startswith("$$"):
            return full[1:]  # strip the leading $

        var_name = match.group("name") or match.group("bare")
        default = match.group("default")

        if var_name is None:
            # Shouldn't happen given the regex, but defensive.
            return full

        env_value = os.environ.get(var_name)
        if env_value is None or env_value == "":
            if default is not None:
                return default
            location = f" in {source_path}" if source_path else ""
            raise InterpolationError(
                f"environment variable ${{{var_name}}} is not set{location}. "
                f"Either export it before running ez-cdc, add it to a .env file, "
                f"or provide a default with ${{{var_name}:-default}}."
            )
        return env_value

    return _VAR_PATTERN.sub(_replace, value)


def _interpolate_recursive(obj: Any, source_path: Path | None) -> Any:
    """Recursively walk a parsed YAML structure interpolating string values."""
    if isinstance(obj, str):
        return interpolate_string(obj, source_path)
    if isinstance(obj, dict):
        return {k: _interpolate_recursive(v, source_path) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_interpolate_recursive(item, source_path) for item in obj]
    return obj


# ── .env file loading ────────────────────────────────────────────────────────

def load_env_files(*paths: Path) -> None:
    """Load one or more `.env` files into os.environ.

    Existing variables are NOT overwritten — the user's shell environment
    always wins. Missing files are silently ignored. Lines that don't parse
    as KEY=VALUE are skipped.

    This is the same minimal parser as in backends/snowflake.py — kept here
    so the loader is self-contained.
    """
    for path in paths:
        if not path.exists():
            continue
        try:
            text = path.read_text()
        except OSError:
            continue

        for raw_line in text.splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" not in line:
                continue
            key, _, value = line.partition("=")
            key = key.strip()
            value = value.strip()
            # Strip surrounding quotes if balanced
            if len(value) >= 2 and value[0] == value[-1] and value[0] in ('"', "'"):
                value = value[1:-1]
            os.environ.setdefault(key, value)


# ── Top-level loader ─────────────────────────────────────────────────────────

def load_datasources(
    path: Path,
    *,
    env_files: list[Path] | None = None,
    allow_missing: bool = False,
) -> DatasourcesFile:
    """Load and validate a datasources YAML file.

    Args:
        path: path to the datasources YAML.
        env_files: optional list of `.env` files to load into os.environ
            BEFORE interpolation. Use this to make `${PROD_PG_PASSWORD}`
            references resolve from a local `.env.prod` file without
            requiring the user to export the variable manually.
        allow_missing: if True and the file doesn't exist, return an empty
            DatasourcesFile instead of raising. Useful for the gate that
            asks "do we have datasources yet?" without failing.

    Returns:
        A DatasourcesFile with the parsed and validated content.

    Raises:
        DatasourceError: if the file doesn't exist (when allow_missing=False),
            can't be read, can't be parsed as YAML, has unresolved `${VAR}`
            references, or fails schema validation.
    """
    if env_files:
        load_env_files(*env_files)

    if not path.exists():
        if allow_missing:
            return DatasourcesFile()
        raise DatasourceError(
            f"datasources file not found: {path}\n"
            f"  Run `ez-cdc datasource init-demos` to create the bundled demo datasources, "
            f"or `ez-cdc datasource add` to configure your own."
        )

    try:
        raw_text = path.read_text()
    except OSError as e:
        raise DatasourceError(f"failed to read {path}: {e}") from e

    try:
        parsed = yaml.safe_load(raw_text) or {}
    except yaml.YAMLError as e:
        # yaml errors include line/column info; surface them as-is.
        raise DatasourceError(f"failed to parse {path} as YAML: {e}") from e

    if not isinstance(parsed, dict):
        raise DatasourceError(
            f"top-level YAML in {path} must be a mapping, got {type(parsed).__name__}"
        )

    # Resolve ${VAR} interpolation across the whole tree before validating.
    try:
        interpolated = _interpolate_recursive(parsed, path)
    except InterpolationError:
        raise  # already has a good message

    # Hand to pydantic.
    try:
        return DatasourcesFile.model_validate(interpolated)
    except ValidationError as e:
        # Format pydantic errors with the YAML field path so the user can find
        # the offending line. Pydantic v2 errors are a list of {loc, msg, type}.
        lines = [f"datasources file {path} failed schema validation:"]
        for err in e.errors():
            loc = ".".join(str(part) for part in err["loc"])
            msg = err["msg"]
            lines.append(f"  • {loc}: {msg}")
        raise DatasourceValidationError("\n".join(lines)) from e

File: e2e-cli/src/config/loader.rs
//! YAML loader and `${VAR}` interpolation for datasources files.
//!
//! Pipeline:
//!   1. Read raw YAML text from disk.
//!   2. Optionally load `.env` files into the process environment.
//!   3. Walk every string value and substitute `${VAR}` with the env var.
//!   4. Deserialize the interpolated tree into `DatasourcesFile` via serde.
//!
//! Interpolation syntax:
//!   `${VAR}`            — required (error if not set)
//!   `${VAR:-default}`   — optional (uses default if unset/empty)
//!   `$$VAR` / `$${VAR}` — escape (literal `$`)
//!   `$VAR`              — bare reference (also supported)

use std::path::Path;

use regex::Regex;
use thiserror::Error;

use super::schema::DatasourcesFile;

// ── Errors ─────────────────────────────────────────────────────────────────

#[derive(Debug, Error)]
pub enum DatasourceError {
    #[error("{0}")]
    NotFound(String),

    #[error("{0}")]
    Validation(String),

    #[error("{0}")]
    Interpolation(String),

    #[error("{0}")]
    Io(String),

    #[error("{0}")]
    Yaml(String),
}

// ── Variable interpolation ─────────────────────────────────────────────────

/// Substitute `${VAR}` placeholders in a single string.
pub fn interpolate_string(value: &str, source_path: Option<&Path>) -> Result<String, DatasourceError> {
    if !value.contains('$') {
        return Ok(value.to_string());
    }

    let re = Regex::new(
        r"(?x)
        \$\$ (?: \{[^}]*\} | [A-Za-z_][A-Za-z0-9_]* )    # escape: $${VAR} or $$VAR
        |
        \$\{ (?P<name>[A-Za-z_][A-Za-z0-9_]*) (?: :- (?P<default>[^}]*) )? \}
        |
        \$ (?P<bare> [A-Za-z_][A-Za-z0-9_]* )              # bare $VAR
        "
    ).expect("valid regex");

    let mut result = String::with_capacity(value.len());
    let mut last_end = 0;

    for caps in re.captures_iter(value) {
        let m = caps.get(0).unwrap();
        result.push_str(&value[last_end..m.start()]);
        last_end = m.end();

        let full = m.as_str();

        // Escape: $${...} or $$VAR → strip one $
        if full.starts_with("$$") {
            result.push_str(&full[1..]);
            continue;
        }

        let var_name = caps.name("name")
            .or_else(|| caps.name("bare"))
            .map(|m| m.as_str());

        let Some(var_name) = var_name else {
            result.push_str(full);
            continue;
        };

        let default = caps.name("default").map(|m| m.as_str());

        match std::env::var(var_name) {
            Ok(v) if !v.is_empty() => result.push_str(&v),
            _ => {
                if let Some(default) = default {
                    result.push_str(default);
                } else {
                    let location = source_path
                        .map(|p| format!(" in {}", p.display()))
                        .unwrap_or_default();
                    return Err(DatasourceError::Interpolation(format!(
                        "environment variable ${{{var_name}}} is not set{location}. \
                         Either export it before running ez-cdc, add it to a .env file, \
                         or provide a default with ${{{var_name}:-default}}."
                    )));
                }
            }
        }
    }
    result.push_str(&value[last_end..]);
    Ok(result)
}

/// Recursively walk a `serde_json::Value` tree, interpolating all strings.
fn interpolate_recursive(
    value: serde_json::Value,
    source_path: Option<&Path>,
) -> Result<serde_json::Value, DatasourceError> {
    match value {
        serde_json::Value::String(s) => {
            let interpolated = interpolate_string(&s, source_path)?;
            Ok(serde_json::Value::String(interpolated))
        }
        serde_json::Value::Object(map) => {
            let mut out = serde_json::Map::new();
            for (k, v) in map {
                out.insert(k, interpolate_recursive(v, source_path)?);
            }
            Ok(serde_json::Value::Object(out))
        }
        serde_json::Value::Array(arr) => {
            let out: Result<Vec<_>, _> = arr
                .into_iter()
                .map(|v| interpolate_recursive(v, source_path))
                .collect();
            Ok(serde_json::Value::Array(out?))
        }
        other => Ok(other),
    }
}

// ── .env file loading ──────────────────────────────────────────────────────

/// Load `.env` files into the process environment.
///
/// Existing variables are NOT overwritten. Missing files are silently ignored.
pub fn load_env_files(paths: &[&Path]) {
    for path in paths {
        let text = match std::fs::read_to_string(path) {
            Ok(t) => t,
            Err(_) => continue,
        };

        for raw_line in text.lines() {
            let line = raw_line.trim();
            if line.is_empty() || line.starts_with('#') {
                continue;
            }
            let Some((key, value)) = line.split_once('=') else {
                continue;
            };
            let key = key.trim();
            let mut value = value.trim().to_string();

            // Strip balanced surrounding quotes.
            if value.len() >= 2 {
                let first = value.as_bytes()[0];
                let last = value.as_bytes()[value.len() - 1];
                if first == last && (first == b'"' || first == b'\'') {
                    value = value[1..value.len() - 1].to_string();
                }
            }

            // setdefault semantics: only set if not already present.
            if std::env::var(key).is_err() {
                std::env::set_var(key, &value);
            }
        }
    }
}

// ── Top-level loader ───────────────────────────────────────────────────────

/// Load and validate a datasources YAML file.
///
/// With `allow_missing=true`, a non-existent file returns an empty
/// `DatasourcesFile` instead of an error.
pub fn load_datasources(
    path: &Path,
    env_files: &[&Path],
    allow_missing: bool,
) -> Result<DatasourcesFile, DatasourceError> {
    if !env_files.is_empty() {
        load_env_files(env_files);
    }

    if !path.exists() {
        if allow_missing {
            return Ok(DatasourcesFile::default());
        }
        return Err(DatasourceError::NotFound(format!(
            "datasources file not found: {}\n  \
             Run `ez-cdc datasource init` to create the bundled demo datasources, \
             or `ez-cdc datasource add` to configure your own.",
            path.display()
        )));
    }

    let raw_text = std::fs::read_to_string(path)
        .map_err(|e| DatasourceError::Io(format!("failed to read {}: {e}", path.display())))?;

    // Parse YAML into a generic JSON value for interpolation.
    let parsed: serde_json::Value = serde_yml::from_str(&raw_text)
        .map_err(|e| DatasourceError::Yaml(format!("failed to parse {} as YAML: {e}", path.display())))?;

    if !parsed.is_object() {
        return Err(DatasourceError::Yaml(format!(
            "top-level YAML in {} must be a mapping, got {}",
            path.display(),
            match &parsed {
                serde_json::Value::Array(_) => "array",
                serde_json::Value::String(_) => "string",
                serde_json::Value::Number(_) => "number",
                serde_json::Value::Bool(_) => "boolean",
                serde_json::Value::Null => "null",
                _ => "unknown",
            }
        )));
    }

    // Interpolate ${VAR} across the whole tree.
    let interpolated = interpolate_recursive(parsed, Some(path))?;

    // Deserialize into the typed schema.
    let file: DatasourcesFile = serde_json::from_value(interpolated)
        .map_err(|e| DatasourceError::Validation(format!(
            "datasources file {} failed schema validation: {e}",
            path.display()
        )))?;

    // Run additional validation (pg ident patterns, table names, etc.).
    file.validate()
        .map_err(|e| DatasourceError::Validation(format!(
            "datasources file {} failed validation: {e}",
            path.display()
        )))?;

    Ok(file)
}

// ── Tests ──────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn interpolate_simple_var() {
        std::env::set_var("EZ_TEST_VAR", "hello");
        let result = interpolate_string("prefix_${EZ_TEST_VAR}_suffix", None).unwrap();
        assert_eq!(result, "prefix_hello_suffix");
        std::env::remove_var("EZ_TEST_VAR");
    }

    #[test]
    fn interpolate_default_value() {
        std::env::remove_var("EZ_MISSING_VAR");
        let result = interpolate_string("val=${EZ_MISSING_VAR:-fallback}", None).unwrap();
        assert_eq!(result, "val=fallback");
    }

    #[test]
    fn interpolate_missing_var_error() {
        std::env::remove_var("EZ_REALLY_MISSING");
        let err = interpolate_string("${EZ_REALLY_MISSING}", None).unwrap_err();
        let msg = err.to_string();
        assert!(msg.contains("EZ_REALLY_MISSING"));
        assert!(msg.contains("is not set"));
    }

    #[test]
    fn interpolate_escape() {
        let result = interpolate_string("$${NOT_A_VAR}", None).unwrap();
        assert_eq!(result, "${NOT_A_VAR}");
    }

    #[test]
    fn interpolate_bare_var() {
        std::env::set_var("EZ_BARE", "works");
        let result = interpolate_string("$EZ_BARE!", None).unwrap();
        assert_eq!(result, "works!");
        std::env::remove_var("EZ_BARE");
    }

    #[test]
    fn interpolate_no_dollar_passthrough() {
        let result = interpolate_string("no vars here", None).unwrap();
        assert_eq!(result, "no vars here");
    }

    #[test]
    fn load_real_yaml() {
        let yaml_path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("e2e/ez-cdc.yaml");
        if yaml_path.exists() {
            let file = load_datasources(&yaml_path, &[], false).unwrap();
            assert!(file.has_any());
            assert!(file.sources.contains_key("demo-pg"));
        }
    }

    #[test]
    fn load_missing_allow() {
        let file = load_datasources(Path::new("/tmp/nonexistent-ez-cdc.yaml"), &[], true).unwrap();
        assert!(file.is_empty());
    }

    #[test]
    fn load_missing_deny() {
        let err = load_datasources(Path::new("/tmp/nonexistent-ez-cdc.yaml"), &[], false).unwrap_err();
        assert!(matches!(err, DatasourceError::NotFound(_)));
    }

    #[test]
    fn load_env_file() {
        let dir = std::env::temp_dir().join("ez-cdc-test-env");
        std::fs::create_dir_all(&dir).unwrap();
        let env_path = dir.join(".env.test");
        {
            let mut f = std::fs::File::create(&env_path).unwrap();
            writeln!(f, "EZ_ENV_TEST_KEY=from_env_file").unwrap();
            writeln!(f, "# comment line").unwrap();
            writeln!(f, "EZ_ENV_QUOTED=\"quoted_val\"").unwrap();
        }

        std::env::remove_var("EZ_ENV_TEST_KEY");
        std::env::remove_var("EZ_ENV_QUOTED");

        load_env_files(&[env_path.as_path()]);

        assert_eq!(std::env::var("EZ_ENV_TEST_KEY").unwrap(), "from_env_file");
        assert_eq!(std::env::var("EZ_ENV_QUOTED").unwrap(), "quoted_val");

        std::env::remove_var("EZ_ENV_TEST_KEY");
        std::env::remove_var("EZ_ENV_QUOTED");
        let _ = std::fs::remove_dir_all(&dir);
    }
}
